wpilibc/src/main/native/cpp/controller/PIDController.cpp

```diff
@@ -0,0 +1,274 @@
+                                        double deltaTolerance) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_toleranceType = Tolerance::kPercent;
+  m_tolerance = tolerance;
+  m_deltaTolerance = deltaTolerance;
+}
+
+double PIDController::GetError() const {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return GetContinuousError(m_reference - m_measurementSource());
+}
+
+/**
+ * Returns the change in error per second of the PIDController.
+ *
+ * @return The change in error per second.
+ */
+double PIDController::GetDeltaError() const {
+  double error = GetError();
+
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return (error - m_prevError) / GetPeriod();
+}
+
+double PIDController::Update() {
+  // Storage for function inputs
+  double Kp;
+  double Ki;
+  double Kd;
+  double feedforward = m_feedforward();
+  double measurement = m_measurementSource();
+  double minimumOutput;
+  double maximumOutput;
+
+  // Storage for function input-outputs
+  double prevError;
+  double error;
+  double totalError;
+
+  {
+    std::lock_guard<wpi::mutex> lock(m_thisMutex);
+
+    Kp = m_Kp;
+    Ki = m_Ki;
+    Kd = m_Kd;
+    minimumOutput = m_minimumOutput;
+    maximumOutput = m_maximumOutput;
+
+    prevError = m_prevError;
+    error = GetContinuousError(m_reference - measurement);
+    totalError = m_totalError;
+  }
+
+  if (Ki != 0) {
+    totalError = clamp(totalError + error * GetPeriod(), minimumOutput / Ki,
+                       maximumOutput / Ki);
+  }
+
+  double output =
+      clamp(Kp * error + Ki * totalError +
+                Kd * (error - prevError) / GetPeriod() + feedforward,
+            minimumOutput, maximumOutput);
+
+  {
+    std::lock_guard<wpi::mutex> lock(m_thisMutex);
+    m_prevError = m_error;
+    m_error = error;
+    m_totalError = totalError;
+    m_output = output;
+  }
+
+  return output;
+}
+
+void PIDController::Reset() {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_prevError = 0;
+  m_totalError = 0;
+  m_output = 0;
+}
+
+void PIDController::InitSendable(SendableBuilder& builder) {
+  builder.SetSmartDashboardType("PIDController");
+  builder.SetSafeState([=]() { Reset(); });
+  builder.AddDoubleProperty("Kp", [=] { return GetP(); },
+                            [=](double value) { SetP(value); });
+  builder.AddDoubleProperty("Ki", [=] { return GetI(); },
+                            [=](double value) { SetI(value); });
+  builder.AddDoubleProperty("Kd", [=] { return GetD(); },
+                            [=](double value) { SetD(value); });
+  builder.AddDoubleProperty(
+      "feedforward", [=] { return m_feedforward(); },
+      [=](double value) { m_feedforward = [=] { return value; }; });
+  builder.AddDoubleProperty("reference", [=] { return GetReference(); },
+                            [=](double value) { SetReference(value); });
+}
+
+double PIDController::GetContinuousError(double error) const {
+  if (m_continuous && m_inputRange != 0) {
```

Comment:
In the Java version, you check `m_inputRange > 0` here. Is the difference here intended?